File: src/ssvep/fbcca-js/filterbank.js
const { decimate, resample, filter, filtfilt, cheb1ord, cheby1 } = require('./fb_helper.js');

function filterbank(eeg, fs_original, idx_fb, target_fs = 256) {
    if (arguments.length < 2) {
        throw new Error('Not enough input arguments.');
    }

    if (arguments.length < 3 || idx_fb === undefined) {
        console.warn('Missing filter index. Default value (idx_fb = 1) will be used.');
        idx_fb = 1;
    } else if (idx_fb < 1 || idx_fb > 10) {
        throw new Error('The number of sub-bands must be 0 < idx_fb <= 10.');
    }

    const num_chans = eeg.length;
    const data_length = eeg[0].length;
    const fs = fs_original / 2;

    // for set of freq 8-15.8Hz (public dataset) as set by default in Nakanishi code
    const passband = [6, 14, 22, 30, 38, 46, 54, 62, 70, 78];
    const stopband = [4, 10, 16, 24, 32, 40, 48, 56, 64, 72];

    // const Wp = [
    //     truncateToDecimalPlaces(passband[idx_fb - 1] / fs, 4), // [idx_fb - 1] is added because MATLAB arrays start from 1
    //     truncateToDecimalPlaces(90 / fs, 4)
    // ];
    const Wp = [passband[idx_fb - 1] / fs, 90 / fs]

    // const Ws = [
    //     truncateToDecimalPlaces(stopband[idx_fb - 1] / fs, 4), // [idx_fb - 1] is added because MATLAB arrays start from 1
    //     truncateToDecimalPlaces(100 / fs, 4)
    // ];

    const Ws = [stopband[idx_fb - 1] / fs, 100 / fs]

    // // for set of freq 8-15.8Hz (public dataset) as determined in paper Chen 2015    
    // const passband = [8, 16, 24, 32, 40]; 
    // const stopband = [4, 10, 18, 26, 34];

    // const Wp = [
    //     truncateToDecimalPlaces(passband[idx_fb - 1] / fs, 4), // [idx_fb - 1] is added because MATLAB arrays start from 1
    //     truncateToDecimalPlaces(88 / fs, 4)
    // ];

    // const Ws = [
    //     truncateToDecimalPlaces(stopband[idx_fb - 1] / fs, 4), // [idx_fb - 1] is added because MATLAB arrays start from 1
    //     truncateToDecimalPlaces(98 / fs, 4)
    // ];    

    // // for Alison's dataset
    // const passband = [4, 12, 20, 28, 36];
    // const stopband = [2, 8, 14, 22, 30];

    // const Wp = [
    //     truncateToDecimalPlaces(passband[idx_fb - 1] / fs, 4), // [idx_fb - 1] is added because MATLAB arrays start from 1
    //     truncateToDecimalPlaces(90 / fs, 4)
    // ];

    // const Ws = [
    //     truncateToDecimalPlaces(stopband[idx_fb - 1] / fs, 4), // [idx_fb - 1] is added because MATLAB arrays start from 1
    //     truncateToDecimalPlaces(100 / fs, 4)
    // ];    

    const { order: N, wn: Wn } = cheb1ord(Wp, Ws, 3, 40);

    const { den: A, num: B } = cheby1(N, 0.5, Wn);

    // Initialize 'y' to the same size as 'eeg'
    let y = new Array(num_chans);
    for (let i = 0; i < num_chans; i++) {
        y[i] = new Array(data_length).fill(0); // Filling with 0s 
    }

    // Assuming B and A are the filter coefficients and 'filter' is a defined function

    for (let ch_i = 0; ch_i < num_chans; ch_i++) {
        // y[ch_i] = filter(B, A, eeg[0][ch_i]);  // Filtering each channel of eeg
        y[ch_i] = filtfilt(B, A, eeg[ch_i]);  // Using filtfilt for larger eeg
    }


    // ---------------------------------------------- CATERING FOR RESAMPLING/DOWN SAMPLING ------------------------------------
    if (fs_original === target_fs) {
        return y;
    }

    // Downsample or resample
    const downsampleFactor = fs_original / target_fs;
    if (Number.isInteger(downsampleFactor)) {
        y = y.map(channel => decimate(channel, downsampleFactor));
    } else {
        y = y.map(channel => resample(channel, fs_original, target_fs));
    }

    return y;

    // ---------------------------------------------- CATERING FOR RESAMPLING/DOWN SAMPLING ------------------------------------
}

module.exports = { filterbank };

File: src/ssvep/fbcca-py/filterbank.py
import numpy as np
from scipy.signal import cheb1ord, cheby1, filtfilt, resample, decimate
from fbcca_config import fbcca_config

# Cache filter coefficients per sub-band so we only design each filter once
_FILTER_COEFF_CACHE = {}

# Sub-band definitions used across calls
_PASSBAND = [6, 14, 22, 30, 38, 46, 54, 62, 70, 78]
_STOPBAND = [4, 10, 16, 24, 32, 40, 48, 56, 64, 72]

def resample_eeg(eeg, original_fs, target_fs=256):
    num_samples = int(eeg.shape[1] * target_fs / original_fs)
    return resample(eeg, num_samples, axis=1)  # Resample along time axis

def downsample_eeg(eeg, original_fs, target_fs=256):
    factor = original_fs // target_fs  # Integer factor
    if original_fs % target_fs != 0:
        raise ValueError("Downsampling factor must be an integer. Use resampling instead.")

    return decimate(eeg, factor, axis=1, ftype='iir')  # Decimate along time axis

def filterbank(eeg, idx_fb=1, target_fs=256):   
    if eeg is None or idx_fb is None:
        raise ValueError('Not enough input arguments.')

    if idx_fb < 1 or idx_fb > 10:
        raise ValueError('The number of sub-bands must be 0 < idx_fb <= 10.')

    num_chans, _ = eeg.shape
    fs_original = fbcca_config.samplingRate
    fs = fs_original / 2

    if idx_fb not in _FILTER_COEFF_CACHE:
        Wp = [_PASSBAND[idx_fb - 1] / fs, 90 / fs]
        Ws = [_STOPBAND[idx_fb - 1] / fs, 100 / fs]
        N, Wn = cheb1ord(Wp, Ws, 3, 40)
        _FILTER_COEFF_CACHE[idx_fb] = cheby1(N, 0.5, Wn, btype='band')

    B, A = _FILTER_COEFF_CACHE[idx_fb]

    # Filter the EEG data
    y = np.zeros_like(eeg)
    
    # Apply the filter to each channel
    for ch_i in range(num_chans):
        y[ch_i, :] = filtfilt(B, A, eeg[ch_i, :], padtype = None)
        
    if fs_original == target_fs:
        return y
    
    # Downsample or resample
    if fs_original % target_fs == 0:
        y = downsample_eeg(y, fs_original, target_fs)
    else:
        y = resample_eeg(y, fs_original, target_fs)

    return y